crates/fuel-core/src/service/genesis/workers.rs

```diff
@@ -0,0 +1,289 @@
+}
+
+impl Handler {
+    fn new(block_height: BlockHeight) -> Self {
+        Self {
+            output_index: 0,
+            block_height,
+        }
+    }
+}
+
+impl HandlesGenesisResource for CoinConfig {
+    fn genesis_resource() -> GenesisResource {
+        GenesisResource::Coins
+    }
+}
+
+impl HandlesGenesisResource for MessageConfig {
+    fn genesis_resource() -> GenesisResource {
+        GenesisResource::Messages
+    }
+}
+
+impl HandlesGenesisResource for ContractConfig {
+    fn genesis_resource() -> GenesisResource {
+        GenesisResource::Contracts
+    }
+}
+
+impl HandlesGenesisResource for ContractStateConfig {
+    fn genesis_resource() -> GenesisResource {
+        GenesisResource::ContractStates
+    }
+}
+
+impl HandlesGenesisResource for ContractBalanceConfig {
+    fn genesis_resource() -> GenesisResource {
+        GenesisResource::ContractBalances
+    }
+}
+
+impl HandlesGenesisResource for ContractId {
+    fn genesis_resource() -> GenesisResource {
+        GenesisResource::ContractsRoot
+    }
+}
+
+impl ProcessState<CoinConfig> for Handler {
+    fn process(&mut self, coin: CoinConfig, tx: &mut Database) -> anyhow::Result<()> {
+        let root = init_coin(tx, &coin, self.output_index, self.block_height)?;
+        tx.add_coin_root(root)?;
+
+        self.output_index = self.output_index
+                .checked_add(1)
+                .expect("The maximum number of UTXOs supported in the genesis configuration has been exceeded.");
+
+        Ok(())
+    }
+}
+
+impl ProcessState<MessageConfig> for Handler {
+    fn process(
+        &mut self,
+        message: MessageConfig,
+        tx: &mut Database,
+    ) -> anyhow::Result<()> {
+        let root = init_da_message(tx, &message)?;
+        tx.add_message_root(root)?;
+        Ok(())
+    }
+}
+
+impl ProcessState<ContractConfig> for Handler {
+    fn process(
+        &mut self,
+        contract: ContractConfig,
+        tx: &mut Database,
+    ) -> anyhow::Result<()> {
+        init_contract(tx, &contract, self.output_index, self.block_height)?;
+        tx.add_contract_id(contract.contract_id)?;
+
+        self.output_index = self.output_index
+                .checked_add(1)
+                .expect("The maximum number of UTXOs supported in the genesis configuration has been exceeded.");
+
+        Ok::<(), anyhow::Error>(())
+    }
+}
+
+impl ProcessStateGroup<ContractStateConfig> for Handler {
+    fn process_group(
+        &mut self,
+        group: Vec<ContractStateConfig>,
+        tx: &mut Database,
+    ) -> anyhow::Result<()> {
+        tx.update_contract_states(group)?;
+        Ok(())
+    }
+}
+
+impl ProcessStateGroup<ContractBalanceConfig> for Handler {
+    fn process_group(
+        &mut self,
+        group: Vec<ContractBalanceConfig>,
+        tx: &mut Database,
+    ) -> anyhow::Result<()> {
+        tx.update_contract_balances(group)?;
+        Ok(())
+    }
+}
+
+impl ProcessState<ContractId> for Handler {
```

Comment:
Maybe instead of implementing the `ProcessState` trait for `ContractId`, we could create a new type(a wrapper around `ContractId`) that will have the same meaning as `ContractBalanceConfig` or `ContractStateConfig>`. In this case it would be expected how `genesis_contract_ids_iter` function is used and it will contain it with this place